### PR TITLE
RPM signing enhancement

### PR DESCRIPTION
**Pull request by yunguowei**
1. Adding all scp8 patchset that we can select algorithm for non-reputation signature, gpg header, RPM file.
2. Enhancment to make smart not install only non-reputation signed RPM (self signed).

### CODE REVIEW

File: templates/default/template.conf [hunk 1]
```diff
@@ -9,3 +9,10 @@ PREFERRED_VERSION_grub-efi-native = "2.00"
 INHERIT += "sign_rpm"
 
 DISTRO_FEATURES_append = " sign_rpm"
+
+# RPM configuration
+PACKAGECONFIG_append_rpm = " openssl"
```
Comment (jiazhang0) [inline templates/default/template.conf]: Why this line is placed here?

File: recipes-devtools/rpm/files/0077-Enable-configuration-for-gpg-digest-algorithm.patch [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
++++ b/macros/macros.in
+@@ -509,8 +509,11 @@ $_arbitrary_tags_tests	Foo:Bar
+ #%__pgp5_check_password_cmd	%{__pgp} \
+ #	pgps +batchmode=on +verbose=0 +armor=off "%{_pgp_name}" -f
+ 
++%_gpg_digest_algo sha1
++
+ %__gpg_sign_cmd			%{__gpg} \
+ 	gpg --batch --no-verbose --no-armor --passphrase-fd 3 --no-secmem-warning \
++        %{?_gpg_digest_algo:--digest-algo %{_gpg_digest_algo}} \
```
Comment (jiazhang0) [inline recipes-devtools/rpm/files/0077-Enable-configuration-for-gpg-digest-algorithm.patch]: You have defined a default value for %_gpg_digest_algo. So it is not necessary to use ?_gpg_digest_algo here. Always passing --digest-algo %{_gpg_digest_algo} looks good enough already.

File: templates/default/template.conf (hunk 1 above)
Comment (yunguowei) [inline templates/default/template.conf]: SCP has dozens of patch depending on ssl, see:

SRC_URI_append = " ${@base_contains('PACKAGECONFIG', 'openssl', '${EXTRA_OPENSSL_SRC_URI}', '', d)}"

# When OpenSSL is selected, we use it as -the- crypto library
PACKAGECONFIG[openssl] = "--with-openssl --with-usecrypto=openssl,--without-openssl,openssl,"

File: recipes-devtools/rpm/files/0077-Enable-configuration-for-gpg-digest-algorithm.patch (hunk 2 above)
Comment (yunguowei) [inline recipes-devtools/rpm/files/0077-Enable-configuration-for-gpg-digest-algorithm.patch]: Yes, but for a develper debuging on the system, he might comment "%_gpg_digest_algo", and gpg_sign_cmd will not return syntax error with existing implemetation.

File: recipes-devtools/python/python-smartpm/0001-pm-verify-package-s-keyid-with-installed.patch [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
++import commands
+ 
+ from smart.util.filetools import setCloseOnExec
+ from smart.sorter import ChangeSetSorter
+@@ -183,6 +184,22 @@ class RPMPackageManager(PackageManager):
+                     if sysconf.get("rpm-check-signatures", True):
+                          if get_public_key(h) == '(none)':
+                              raise rpm.error('package is not signed')
++                         else:
++                              (cstatus, coutput) = commands.getstatusoutput("rpm -Kv %s |grep 'signature: OK'|grep 'key ID'" %(path))
```
Comment (jiazhang0) [inline recipes-devtools/python/python-smartpm/0001-pm-verify-package-s-keyid-with-installed.patch]: We can only search 'key ID' here, right? 

"|grep" needs a space for separating.

File: recipes-devtools/python/python-smartpm/0001-pm-verify-package-s-keyid-with-installed.patch [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
++                              idx = coutput.find('key ID')+7
```
Comment (jiazhang0) [inline recipes-devtools/python/python-smartpm/0001-pm-verify-package-s-keyid-with-installed.patch]: Use 'key ID ' here.

File: recipes-devtools/python/python-smartpm/0001-pm-verify-package-s-keyid-with-installed.patch [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
++                              keyid = coutput[idx:].strip()
```
Comment (jiazhang0) [inline recipes-devtools/python/python-smartpm/0001-pm-verify-package-s-keyid-with-installed.patch]: If "NOKEY" appears in the line, keyid here is "153804d9: NOKEY", and the compare below will fail.

Comment (jiazhang0) [inline recipes-devtools/python/python-smartpm/0001-pm-verify-package-s-keyid-with-installed.patch]: Or is it impossible "NOKEY" case reaches to here?

Comment (yunguowei) [inline recipes-devtools/python/python-smartpm/0001-pm-verify-package-s-keyid-with-installed.patch]: No, NOKEY means signature check failed.

File: recipes-devtools/python/python-smartpm/0001-pm-verify-package-s-keyid-with-installed.patch (hunk 3 above)
Comment (yunguowei) [inline recipes-devtools/python/python-smartpm/0001-pm-verify-package-s-keyid-with-installed.patch]: Done

File: recipes-devtools/python/python-smartpm/0001-pm-verify-package-s-keyid-with-installed.patch (hunk 4 above)
Comment (yunguowei) [inline recipes-devtools/python/python-smartpm/0001-pm-verify-package-s-keyid-with-installed.patch]: Done

**Comment by jiazhang0** [comment]
As discussed, shall we move all definitions in the default templates to a new template called rpm-signing?

**Comment by yunguowei** [comment]
I'm not sure if I have effort to do that shortly, so I would like to have these patch merged if you don't mind, as it can avoid context conflict with other commits. I can submit additional commit later to reshape rpm-signing template.

**Comment by jiazhang0** [comment]
The patchset itself is good. Just need to a cleanup in future.

**Pull request merged**